test(QDialog): add QDialog tests

n05la3 opened the pull request:
<!--
  Please make sure to read the Pull Request Guidelines:
  https://github.com/quasarframework/quasar/blob/dev/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- Update "[ ]" to "[x]" to check a box (space sensitive) -->

**What kind of change does this PR introduce?** <!-- Check at least one -->

- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [x] Other, please describe:

**Does this PR introduce a breaking change?** <!-- Check one -->

- [ ] Yes
- [x] No

<!-- If yes, please describe the impact and migration path for existing applications: -->

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch (or `v[X]` branch)
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on a Cordova (iOS, Android) app
- [ ] It's been tested on an Electron app
- [ ] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar/tree/dev/docs) <!-- for faster update click on "Suggest an edit on GitHub" at bottom of page --> or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to [start a new feature discussion](https://github.com/quasarframework/quasar/discussions/new?category=ideas-proposals) first and wait for approval before working on it)

**Other information:**

n05la3 committed:
test(QDialog): add QDialog tests

IlCallo commented on ui/src/components/dialog/__tests__/DialogWrapper.vue:
```diff
@@ -0,0 +1,43 @@
+    const dialogRef = ref(null)
+
+    function focus () {
+      console.log(dialogRef.value.focus, 'Attempting to focus')
```
I guess this is a leftover

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,82 @@
+import FieldWrapper from './DialogWrapper.vue'
```
Wrong import name, should be `DialogWrapper`

IlCallo commented on ui/src/components/dialog/__tests__/DialogWrapper.vue:
```diff
@@ -0,0 +1,43 @@
+      </q-card>
+    </q-dialog>
+
+    <q-btn data-cy="close-button" label="Open Dialog"/>
```
I guess you're not really using this one, since it's not referenced and the label and data-cy are opposites

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,82 @@
+          })
+
+          cy.withinDialog(() => {
+            closeDialog() // Closes dialog by clicking backdrop
```
I'd just rename the function as `closeDialogViaBackdrop`

IlCallo replied:
This test should check that the Dialog cannot be closed either via ESC, nor backdrop, nor route navigation

I'd also avoid testing the `persistent: false` behaviour, since that will be tested in the other 3 tests about the specific props (`no-esc-dismiss`, etc)

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,82 @@
+          })
+
+          getHostElement().then(() => {
+            getHostElement().trigger('keydown', { keyCode: 27 })
```
Shouldn't this be `cy.get('body').type('{esc}')` too?
Why using keydown here?

n05la3 replied:
Fixed.

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,82 @@
+          })
+
+          cy.withinDialog(() => {
+            closeDialog()
```
Using it here is an even harder hint that this should be named `closeDialogViaBackdrop`

IlCallo reviewed (changes requested):
Really nice work!
Aside some dead code around and a couple readability fixes, everything seems fine

n05la3 committed:
test(QDialog): improve test stability

n05la3 committed:
test(use-model-toggle): improve test stability

n05la3 commented:
I have resolved the issues with the PR. 

As concerns the `use-model-toggle`, it was a race condition that caused the issues. After digging, I noticed inconsistencies between the tests and the code when testing the `hide` event. The hide function in `QMenu.js` looks something like this:
```ts
function handleHide (evt) {
  hidePortal() // The dialog is hidden

  // Some code here. It's how fast this happens that determines if the test will pass or fail

  emit('hide', evt)
}
```
While in the test looks something like this:
```ts
cy.dataCy('method-hide')
  .click({ force: true })
cy.dataCy('menu')
  .should('not.exist')
  .then(() => {
    expect(fn).to.be.called
  })
```
- First, there's no guarantee that the click is completed before checking if the menu no longer exists, this therefore creates a race condition and depends on how fast the hide method resolves. Normally, it should check that the menu existed before running the hide method. Otherwise, it will just depend on how fast things happen to pass or fail. In addition to this problem:
- Checking if menu no longer exists is not the right thing to do because the `hidePortal` gets called the `hide` event is emitted. So, the menu may disappear before the event is emitted. This should have been passing because things were happening so fast.

With the `QDialog`, it mainly needed to ensure that the dialog existed before continuing the tests. A few tests were flaky too and only passed depending on how fast things were happening. We had tests like this:

```ts
 cy.dataCy('dialog-page').then(() => {
 1)   cy.dataCy('input-field').focus()
 2)  model.value = true
  
    cy.withinDialog(() => {
      closeDialog()
    })
    cy.dataCy('input-field').should('have.focus')
})
 ```
It has a similar problem to the `q-toggle-model`. When run (1), we are not sure whether or not it succeeds. For instance, in some cases, after running (2), the backdrop will cover the page before completing the `focus`. So (1) may silently fail and the tests after that are false positives. Instead, this has been replaced with the more stable counterpart below:
```ts
cy.dataCy('input-field')
    .type('Hello')
  cy.dataCy('input-field')
    .should('have.value', 'Hello').then(() => {
      model.value = true

      cy.withinDialog(() => {
        closeDialogViaBackdrop()
      })
      getHostElement()
        .should('not.exist').then(() => {
          cy.dataCy('input-field')
            .should('have.focus')
        })
    })
```
I have made similar changes to the rest of `QDialog` tests to improve stability.

IlCallo committed:
test(QDialog): refactor and fix some tests

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,131 @@
+  return cy.get('.q-dialog__backdrop').click({ force: true })
+}
+
+// TODO: this only works in `withinDialog` context,
```
Check out this TODO
If you find a way to make `closeDialogViaEscKey` work both inside and outside a `within` context, that would be cool

I tried with `cy.window()` and `cy.document()` already, which I recalled worked fine, but it seems like they changed that

n05la3 replied:
Could not find a better way to handle this either.

IlCallo replied:
Ok, let's keep the TODO and see if we are able to spend some more time on this in a follow up PR

IlCallo replied:
Documented these new helpers and their limitations here: https://github.com/quasarframework/quasar-testing/issues/355

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,131 @@
+  return cy.root().closest('body').type('{esc}')
+}
+
+// TODO: we'll add this check to the AE helper to automatically test persistent dialogs
```
This is okeish, but it doesn't account for seamless dialogs and dialogs without a backdrop
If we can find some other way to test for persistent dialogs being there, we can execute it automatically whenever a `withinDialog` command is called with `persistent: true`

n05la3 replied:
Checked the DOM again with these props and could not find any better way to identify persistent dialogs. We could probably add `cy.get('.q-dialog').should('not.have.class', 'q-dialog--seamless')` too but it is no guarantee that the persistent prop was set. So I left it as it is.

IlCallo replied:
Ok, we can't add this automatically into the Cypress AE helper then, but we can document this pattern and its limitations into the AE README, until we find a proper solution

IlCallo replied:
Documented these new helpers and their limitations here: https://github.com/quasarframework/quasar-testing/issues/355

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,131 @@
+
+          cy.wrap()
+            .then(async () => {
+              // Care as you need to wait setProps to complete its execution
```
`Cypress.vueWrapper.setProps()` is an async function
If you just call it, without marking the `then` callback as `async`, Cypress wouldn't await for it to complete before proceeding with other commands, possibly breaking the test or causing a race condition

Check out if VTU or Cypress already provide an helper which minimize this category of errors
If not, you can try to create your own which automate the `cy.wrap().then(async () => {})` bit, even if the user would then still need to manually await its promises

If it's generic enough, we can add it to Cypress AE later on

n05la3 replied:
What do you think about this?
```ts
function getPreviousElement (fn) {
  return cy.wrap()
    .then(async () => {
      await fn()
    })
}
```
I wrote something like this but did not find any real benefit to using it instead of just calling `cy.wrap`. Even writing it as `cy.getPrevious()` will still end up with a similar amount of typing except it is more meaningful. It may be interesting to have this in the AE, but I don't think it's worth it here. What do you think?

IlCallo replied:
The name feels strange, why using "previous" or "previous element"?
`cy.wrap` doesn't actually touch in any way anything "previous" or an element

n05la3 replied:
Agreed. `cy.getPrevious` is not a good name here. Perhaps something like `getFulfilled`, `wrapAndFulFill`, `wrapAndThen` or something better in this case should be a better name. However, like I said in my previous comment, I don't think it serves some real benefit adding it here. However, I think it could be good in the AE. What do you think?

IlCallo replied:
If you didn't find it useful and we're not using it at least in our own tests, then let's avoid adding it in the AE for now

Let's try to add only what we know is actually useful

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -1,21 +1,131 @@
+              await Cypress.vueWrapper.setProps({ noEscDismiss: false })
+            })
+
+          cy.withinDialog(() => {
```
Try to use `cy.withinDialog` whenever possible, as it automates some checks about Dialog existence and not existence for you

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -26,103 +136,425 @@ describe('Dialog API', () => {
+          cy.dataCy('input-field')
+            .then(async () => {
+              model.value = true
+              await nextTick()
```
I'm not sure `nextTick` is actually required when changing the refs, to make sure there's no race condition, check it out / test it / search for confirmation about this (I guess on VTU repo)

n05la3 replied:
In this particular case, `nextTick` is required. According to VTU docs, if we await `setProps`, `setValue`, etc then it is equivalent to having called the `await Vue.nextTick()` after the command, just as you noted in one of your comments.  I noticed that the `vModelAdapter` implementation does not await when calling `Cypress.vueWrapper.setProps({ [modelName]: value })`. So in this case we need `nextTick` to ensure the view has updated.

IlCallo replied:
Shall we update `vModelAdapter` to await for `setProps` then?
I'm not sure the interaction would work as expected, due to the use of reactivity we do in tests, but we could try that

n05la3 replied:
Yes, we should update `vModelAdapter` and test that it does not break any existing tests. I do not foresee it having any particular issues with existing tests. I'll update and check this out.

IlCallo replied:
Go on and test this

n05la3 replied:
Tested the edit and it does not break any tests. I have opened a PR [here](https://github.com/quasarframework/quasar-testing/pull/364)

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -26,103 +136,425 @@ describe('Dialog API', () => {
+            }
+          })
+
+          cy.dataCy('dialog-form')
```
Could this be done using `withinDialog` and `cy.root()` within it?
Not sure, check it out

In general, it's not totally clear to me how we should deal with seamless dialogs

n05la3 replied:
I ended up using `cy.root()` within `withinDialog`

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -26,103 +136,425 @@ describe('Dialog API', () => {
         })
       })
 
       describe('(prop): maximized', () => {
```
I stopped my refactoring here, continue and re-check out again the model-toggle composable changes

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -26,103 +136,425 @@ describe('Dialog API', () => {
+            }
+          })
+
+          cy.dataCy('dialog-page').then(() => {
```
Using `.get` with `dialog-page` shouldn't be needed, as you're assured for it to exist after mount is executed

IlCallo commented on ui/src/components/dialog/__tests__/QDialog.cy.js:
```diff
@@ -26,103 +136,425 @@ describe('Dialog API', () => {
+            getHostElement().then(() => {
+              cy.dataCy('dialog-form')
+                .get('.q-dialog__inner--maximized')
+                .should('exist')
```
Avoid `should('exist')`, if not to convey an explicit message
Rely on `withinDialog` whenever possible, or on the implicit check hidden in `.get` commands

IlCallo commented on ui/src/composables/private/__tests__/use-model-toggle.cy.js:
```diff
@@ -165,39 +165,53 @@ describe('use-model-toggle API', () => {
-          .should('not.exist')
-          .then(() => {
-            expect(fn).to.be.called
+          .should('exist').then(() => {
```
Remove `.should('exist')` where you added it, as it's implicit in `.dataCy` which is based upon `.get`

IlCallo commented on ui/src/composables/private/__tests__/use-model-toggle.cy.js:
```diff
@@ -165,39 +165,53 @@ describe('use-model-toggle API', () => {
-            expect(fn).to.be.called
+          .should('exist').then(() => {
+            cy.get('body')
+              .click(499, 0)
```
If possible, add a comment explaining these numbers

IlCallo commented on ui/src/components/dialog/__tests__/DialogWrapper.vue:
```diff
@@ -0,0 +1,41 @@
+<template>
+  <div data-cy="dialog-page">
+    <q-dialog ref="dialogRef" v-bind="$attrs" data-cy="dialog-form">
```
`dialog-form` doesn't seem a good name for this
`dialog-container` maybe? Or just `dialog`? Or whatever

If you can avoid referencing this altogether in tests, that would be even better

IlCallo reviewed (changes requested):
I left some comments to highlight some changes I did and explain to which pitfalls you need to pay attention

I didn't complete the refactor, but I marked with a comment up until where I did it
Good luck

n05la3 committed:
test(QDialog): refactor and improve tests

IlCallo committed:
chore(ui) upgrade test deps

IlCallo committed:
test(QDialog): refactor some tests to be more readable

n05la3 committed:
test(use-model-toggle): revert changes on use-model-toggle and prefer stopping animation on QMenu to fix event issues

- After a dialog has been closed, the hide event is fired only after the duration transitionDuration has elapsed. transitionDuration is a prop passed to use-model-toggle.js and has a default value of 300ms. Hence, we cannot depend on the dialog closing to conclude that the event has been fired. Setting transitionDuration to 0 ensures it fires immediately

n05la3 committed:
test(use-model-toggle): add test for 'hide' event interaction with transition duration prop

IlCallo committed:
test(QDialog): apply vModelAdapter fix
